In [ ]:
import numpy as np
import pandas as pd
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


##데이터 불러오기
- 장르유형 '댄스'만 사용


In [ ]:
data = pd.read_pickle('/content/drive/MyDrive/5조_파이널PJT/코드/data/melon.pkl')
data

pages   rank                                              title  \
0          1      1                                             사랑인가 봐   
1          1      2                                               취중고백   
2          1      3                                             듣고 싶을까   
3          1      4                                          사랑은 늘 도망가   
4          1      5                                            언제나 사랑해   
...      ...    ...                                                ...   
94608  26101  26146                              음주운전 하지마 (하지마송part.1)   
94609  26101  26147  All I Want for Christmas Is You (Feat. Edward ...   
94610  26101  26148                                  Starlight (inst.)   
94611  26101  26149                                        사막별 (inst.)   
94612  26101  26150                                    힘내라 친구야 (Inst.)   

                                                   lyric  \
0      너와 함께 하고 싶은 일들을\n상상하는 게\n요즘 내 일상이 되고\n너의 즐거워하는...   
1      뭐하고 있었니 늦었지만\n잠시 나올래\n너의 집 골목에 있는\n놀이터에 앉아 있어\...   
2      우리 함께 듣던 그 노랠 듣고 싶을까\n듣고 나서 잠시 날 생각은 할까\n아주 혹시...   
3      눈물이 난다 이 길을 걸으면\n그 사람 손길이 자꾸 생각이 난다\n붙잡지 못하고 가...   
4      둘이 손잡고 걸을 때\n마주 보며 또 웃을 때\n사랑한다 말하지 않아도 알아\n널 ...   
...                                                  ...   
94608  하지마 하지마 음주운전 하지마\n하지마 하지마 음주운전 하지마\n술먹고 운전했었어\...   
94609  Merry christmas\n큰마음 먹고 널 위해 준비한 데이트\n나 홀로 집에 ...   
94610  저 떨어지는 별에 빌었어\n지금 네 맘도 나와 똑같기를\n잠든 밤이 지나고 네가 떠...   
94611  소중한 무언가를 난 또 쫓고 있어\n어떤 모양인지조차\n생각해 볼 순간도 없고\n너...   
94612  마시자 잔을 들어 친구야\n사느라고 고생이 많다\n막힌 현실에 갇혀 살아도\n잘 버...   

                                artist      genres                 writer  
0                                 멜로망스  발라드, 국내드라마             김민석 (멜로망스)  
1                           김민석 (멜로망스)         발라드                    김희탐  
2                        MSG워너비(M.O.M)         발라드                  Dr.JO  
3                                  임영웅  발라드, 국내드라마                    강태규  
4                          케이시 (Kassy)         발라드                    조영수  
...                                ...         ...                    ...  
94608                               강건          댄스                     강건  
94609  커플다이어리, 아우라 (AOORA), $milli, 윤제    댄스, 인디음악            아우라 (AOORA)  
94610                             ENOi          댄스                  JYMON  
94611                             ENOi          댄스  페이퍼플래닛 (Paper Planet)  
94612                        코드 (KOHD)          댄스                    유해준  

[94613 rows x 7 columns]

In [ ]:
dance = data[data['genres'].str.contains('댄스', na=False)]
dance

pages   rank                                              title  \
1729    1701   1733                                             머피의 법칙   
3533    3501   3539                                  누난 너무 예뻐 (Replay)   
9178    9201   9208                             너 아니면 안되는 걸 (Romantic)   
10882  10901  10933                                     화장을 하고 (Graze)   
14238  14301  14342                              내 곁에만 있어 (Best Place)   
...      ...    ...                                                ...   
94608  26101  26146                              음주운전 하지마 (하지마송part.1)   
94609  26101  26147  All I Want for Christmas Is You (Feat. Edward ...   
94610  26101  26148                                  Starlight (inst.)   
94611  26101  26149                                        사막별 (inst.)   
94612  26101  26150                                    힘내라 친구야 (Inst.)   

                                                   lyric  \
1729   돈 싫어 명예 싫어\n따분한 음악 우린 정말 싫어\n펑키 비트의 신나는 댄스\n노래...   
3533   누난 너무 예뻐서 남자들이\n가만 안 둬\n흔들리는 그녀의 맘 사실\n알고 있어\n...   
9178   Yeah my girl\n내가 바보야\n잊을 수 없는 너의 사랑이\n마지막 눈물이\...   
10882  You gave me\nyou gave me a heart break\ncome o...   
14238  [종현]\n이젠 그만 내가 싫어졌단 말\n거짓말인 걸 알아\n([민호] 다 거짓말이...   
...                                                  ...   
94608  하지마 하지마 음주운전 하지마\n하지마 하지마 음주운전 하지마\n술먹고 운전했었어\...   
94609  Merry christmas\n큰마음 먹고 널 위해 준비한 데이트\n나 홀로 집에 ...   
94610  저 떨어지는 별에 빌었어\n지금 네 맘도 나와 똑같기를\n잠든 밤이 지나고 네가 떠...   
94611  소중한 무언가를 난 또 쫓고 있어\n어떤 모양인지조차\n생각해 볼 순간도 없고\n너...   
94612  마시자 잔을 들어 친구야\n사느라고 고생이 많다\n막힌 현실에 갇혀 살아도\n잘 버...   

                                artist         genres                 writer  
1729                            DJ DOC  발라드, 댄스, 랩/힙합              강은경,  이승호  
3533                      SHINee (샤이니)        발라드, 댄스            YOUNG H.KIM  
9178                      SHINee (샤이니)        발라드, 댄스                    유영진  
10882                     SHINee (샤이니)        발라드, 댄스                    김정배  
14238                     SHINee (샤이니)        발라드, 댄스                     홍석  
...                                ...            ...                    ...  
94608                               강건             댄스                     강건  
94609  커플다이어리, 아우라 (AOORA), $milli, 윤제       댄스, 인디음악            아우라 (AOORA)  
94610                             ENOi             댄스                  JYMON  
94611                             ENOi             댄스  페이퍼플래닛 (Paper Planet)  
94612                        코드 (KOHD)             댄스                    유해준  

[23663 rows x 7 columns]

In [ ]:
dance = dance.head(500)

# 훈련데이터 생성
- 줄바꿈 문자 기준으로 문장 토큰화

In [ ]:
dance_text = dance['lyric'].str.split('\n')
dance_text

1729     [돈 싫어 명예 싫어, 따분한 음악 우린 정말 싫어, 펑키 비트의 신나는 댄스, 노...
3533     [누난 너무 예뻐서 남자들이, 가만 안 둬, 흔들리는 그녀의 맘 사실, 알고 있어,...
9178     [Yeah my girl, 내가 바보야, 잊을 수 없는 너의 사랑이, 마지막 눈물이...
10882    [You gave me, you gave me a heart break, come ...
14238    [[종현], 이젠 그만 내가 싫어졌단 말, 거짓말인 걸 알아, ([민호] 다 거짓말...
                               ...                        
71445    [왜 나만 이래, 매일 밤 혼자, 꼭 반쯤 미친 애처럼 말야, 달이 떠오를 때쯤 I...
71446    [지금 이 순간 후회 없이, 보여주고 싶어, 그저 네 심장을 뛰게 하고 싶어, 눈부...
71447    [나 어떡해요 언니, 내 말을 들어봐, 내 그 사람을 언니, 모르겠어요, 참 엉뚱하...
71448    [왠지 설레는 느낌, Make me feel so high, 환한 미소가, I re...
71449    [푸른 달은 이미 다 지고, 눈을 가린 안개 더 짙어져, 손끝에 집중해 더 깊이, ...
Name: lyric, Length: 500, dtype: object

- 토큰나이저 시행

In [ ]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(dance_text)
vocab_size = len(tokenizer.word_index) + 1
print('단어 집합의 크기 : %d' % vocab_size)

단어 집합의 크기 : 19044


In [ ]:
sequences = list()
for line in dance_text: # 줄바꿈 문자를 기준으로 문장 토큰화
    encoded = tokenizer.texts_to_sequences([line])[0]
    for i in range(1, len(encoded)):
        sequence = encoded[:i+1]
        sequences.append(sequence)

print('학습에 사용할 샘플의 개수: %d' % len(sequences))

학습에 사용할 샘플의 개수: 32430


In [ ]:
max_len = max(len(l) for l in sequences) # 모든 샘플에서 길이가 가장 긴 샘플의 길이 출력
print('샘플의 최대 길이 : {}'.format(max_len))

샘플의 최대 길이 : 217


- pad_sequences()를 사용해 모든 샘플의 길이 맞춰주기
- padding의 'pre'는 0값의 위치 설정 인자
 * 이 경우 샘플 최대 길이인 217보다 짧은 샘플의 앞에 0을 채워 전체 훈련데이터의 길이를 맞춤

In [ ]:
sequences = pad_sequences(sequences, maxlen=max_len, padding='pre')

In [ ]:
sequences = np.array(sequences)
X = sequences[:,:-1]
y = sequences[:,-1]

- 원-핫 인코딩 수행

In [ ]:
y = to_categorical(y, num_classes=vocab_size)

## 모델 설계

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Dense, SimpleRNN

- 하이퍼파라미터 설명
 * 임베딩 벡터 차원 : 10
 * 은닉상태 크기 : 32
 * 다 대 일 구조의 RNN 사용

- 해당 문제 해결 방법 : 다중 클래스 분류 문제
 * 마지막 시점에 모든 가능한 단어 중 하나의 단어를 예측하는 문제로 접근
 * 활성화 함수: 소프트맥스 함수
 * 손실 함수: 크로스 엔트로피 함수
 * 에포크 : 200

In [ ]:
embedding_dim = 10
hidden_units = 32

model = Sequential()
model.add(Embedding(vocab_size, embedding_dim))
model.add(SimpleRNN(hidden_units))
model.add(Dense(vocab_size, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X, y, epochs=5, verbose=2)

Epoch 1/5
1014/1014 - 177s - loss: 9.6115 - accuracy: 0.0512 - 177s/epoch - 174ms/step
Epoch 2/5
1014/1014 - 176s - loss: 9.0585 - accuracy: 0.0539 - 176s/epoch - 173ms/step
Epoch 3/5
1014/1014 - 176s - loss: 8.6424 - accuracy: 0.0546 - 176s/epoch - 173ms/step
Epoch 4/5
1014/1014 - 174s - loss: 8.2429 - accuracy: 0.0550 - 174s/epoch - 172ms/step
Epoch 5/5
1014/1014 - 175s - loss: 7.8809 - accuracy: 0.0580 - 175s/epoch - 172ms/step


# 문장 생성 함수

In [ ]:
def sentence_generation(model, tokenizer, current_word, n): # 모델, 토크나이저, 현재 단어, 반복할 횟수
    init_word = current_word
    sentence = ''

    # n번 반복
    for _ in range(n):
        # 현재 단어에 대한 정수 인코딩과 패딩
        encoded = tokenizer.texts_to_sequences([current_word])[0]
        encoded = pad_sequences([encoded], maxlen=5, padding='pre')
        # 입력한 X(현재 단어)에 대해서 Y를 예측하고 Y(예측한 단어)를 result에 저장.
        result = model.predict(encoded, verbose=0)
        result = np.argmax(result, axis=1)

        for word, index in tokenizer.word_index.items(): 
            # 만약 예측한 단어와 인덱스와 동일한 단어가 있다면 break
            if index == result:
                break

        # 현재 단어 + ' ' + 예측 단어를 현재 단어로 변경
        current_word = current_word + ' '  + word

        # 예측 단어를 문장에 저장
        sentence = sentence + ' ' + word

    sentence = init_word + sentence
    return sentence

파라미터 설명
- sentence_generation(model, tokenizer, '예측에 사용할 단어, 예측 횟수)

In [ ]:
print(sentence_generation(model, tokenizer, '사랑', 4))

사랑    


In [ ]:
test2 = '내가 그의 이름을 불러주기 전에는'

print(sentence_generation(model, tokenizer, test2, 4))

내가 그의 이름을 불러주기 전에는    
